crates/wysiwyg/src/dom/insert_parent.rs

```diff
@@ -0,0 +1,271 @@
+use crate::{DomHandle, DomNode, UnicodeString};
+
+use super::{Dom, Range};
+
+impl<S> Dom<S>
+where
+    S: UnicodeString,
+{
+    #[allow(dead_code)]
+    /// Insert a node and make this node the parent of a given range.
+    pub fn insert_parent(&mut self, range: &Range, mut new_node: DomNode<S>) {
+        #[cfg(any(test, feature = "assert-invariants"))]
+        self.assert_invariants();
+
+        if range.is_empty() {
+            panic!("Cannot add a parent to an empty range");
+        }
+
+        if !matches!(new_node, DomNode::Container(_)) {
+            panic!("New node is not a container");
+        }
+        // Check if the range has shared ancestry. The new parent node should not contain
+        // these nodes, so filter them from the range.
+        let shared_depth = range.shared_parent().depth();
+        let range = Range::new(range.locations_from_depth(shared_depth));
+
+        // Prepare the new parent node to have the selected range moved into it:
+        // - Set the new node's handle to the start of the range
+        // - Get the mutable container node to move the selected range into
+        let new_handle: DomHandle = range.node_handles().min().unwrap().clone();
+        let new_handle = if range.leaves().min().unwrap().is_covered() {
```

Comment:
You won't believe how long it took me to find a good way to decide where the new parent should be inserted when the tree is split (I can't exactly use range with block nodes), so good job here!